### PR TITLE
Randomise etcd backup schedule per shoot

### PR DESCRIPTION
**Pull request by swapnilgm**
Signed-off-by: Swapnil Mhamane <swapnil.mhamane@sap.com>

**What this PR does / why we need it**:
This PR will randomize the schedule for full snapshots
- existing cluster will still have same schedule
- if schedule is configured in provider extension config, the that will be used
- otherwise random schedule within last hour before shoot maintenance window will be chosen.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/etcd-backup-restore/issues/140

**Special notes for your reviewer**:

**Release note**:

```improvement operator
The schedule for full snapshot for etcd backing shoot cluster will be randomised for new shoots. The default schedule will still be configurable via provider extension config, But if not specified it will randomised within last hour before shoot maintenance. If schedule is already set or maintenance window change later on, it won't change the backup schedule
```

### CODE REVIEW

File: pkg/webhook/controlplane/etcd.go [hunk 1]
```diff
@@ -134,3 +144,61 @@ func GetETCDVolumeClaimTemplate(name string, storageClassName *string, storageCa
...
+		return schedule, nil
+	}
+
+	// If default schedule is configured, then we use it.
+	if defaultSchedule != nil && len(*defaultSchedule) > 0 {
+		return *defaultSchedule, nil
+	}
+
+	if cluster.Shoot.Spec.Maintenance != nil && cluster.Shoot.Spec.Maintenance.TimeWindow != nil {
+
```
Comment (rfranzke) [inline pkg/webhook/controlplane/etcd.go]: remove empty line

File: pkg/webhook/controlplane/etcd.go [hunk 2]
```diff
@@ -134,3 +144,61 @@ func GetETCDVolumeClaimTemplate(name string, storageClassName *string, storageCa
...
+	if len(schedule) != 0 {
+		return schedule, nil
+	}
+
+	// If default schedule is configured, then we use it.
+	if defaultSchedule != nil && len(*defaultSchedule) > 0 {
+		return *defaultSchedule, nil
+	}
+
+	if cluster.Shoot.Spec.Maintenance != nil && cluster.Shoot.Spec.Maintenance.TimeWindow != nil {
```
Comment (rfranzke) [inline pkg/webhook/controlplane/etcd.go]: FYI: There is https://github.com/gardener/gardener-extensions/pull/319 which will probably be merged first, and you should then adapt this part to be capable of working with both `cluster.Shoot` and `cluster.CoreShoot`.

Comment (swapnilgm) [reply pkg/webhook/controlplane/etcd.go]: Cool. Thanks for prior info.

**Comment by swapnilgm** [comment]
/needs/review
/ok-to-merge

**Comment by rfranzke** [comment]
Great PR, nice work!

**Comment by ialidzhikov** [comment]
Hey, my etcd locally gets recreated at each reconciliation because of a change in the `--schedule` param. Could you please check this.

/cc @swapnilgm @rfranzke

**Comment by swapnilgm** [comment]
> Hey, my etcd locally gets recreated at each reconciliation because of a change in the --schedule param. Could you please check this.

Accepted. Thanks. The units test doesn't cover this scenario.
RCA: On gardener reconciliation, gardener etcd sts update request comes without backup sidecar config.  Hence the while parsing schedule from existing sts the webhook dosen't find the backup sidecar, hence already assigned schedule and hence try to assign it again using random function which is not constant across multiple iterations.

Will fix it.

**Comment by ialidzhikov** [comment]
That was fast. Thank you! 🚀

**Pull request merged**